NPEs in 'AutomaticInboxProcessor' during startup

```
2019-05-27 12:15:09.364 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'EventSubscriber.receive()' on 'org.eclipse.smarthome.config.discovery.internal.AutomaticInboxProcessor@36897ddd': null
java.lang.NullPointerException: null
	at org.eclipse.smarthome.config.discovery.internal.AutomaticInboxProcessor.receiveTypedEvent(AutomaticInboxProcessor.java:113) ~[?:?]
	at org.eclipse.smarthome.config.discovery.internal.AutomaticInboxProcessor.receiveTypedEvent(AutomaticInboxProcessor.java:1) ~[?:?]
	at org.eclipse.smarthome.core.events.AbstractTypedEventSubscriber.receive(AbstractTypedEventSubscriber.java:54) ~[133:org.openhab.core:2.5.0.201905250301]
	at sun.reflect.GeneratedMethodAccessor38.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:153) [133:org.openhab.core:2.5.0.201905250301]
	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:53) [133:org.openhab.core:2.5.0.201905250301]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```

```
2019-05-27 12:15:09.388 [ERROR] [me.core.internal.events.EventHandler] - Dispatching/filtering event for subscriber 'org.eclipse.smarthome.core.events.EventSubscriber' failed: null
java.lang.NullPointerException: null
	at org.eclipse.smarthome.config.discovery.internal.AutomaticInboxProcessor.receiveTypedEvent(AutomaticInboxProcessor.java:113) ~[?:?]
	at org.eclipse.smarthome.config.discovery.internal.AutomaticInboxProcessor.receiveTypedEvent(AutomaticInboxProcessor.java:1) ~[?:?]
	at org.eclipse.smarthome.core.events.AbstractTypedEventSubscriber.receive(AbstractTypedEventSubscriber.java:54) ~[133:org.openhab.core:2.5.0.201905250301]
	at sun.reflect.GeneratedMethodAccessor38.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:153) [133:org.openhab.core:2.5.0.201905250301]
	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:53) [133:org.openhab.core:2.5.0.201905250301]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```

It looks like the `thingRegistry` has not been initialized / set yet. Could be solved by switching to contructor injection.

https://github.com/openhab/openhab-core/blob/36ae91856c01f65ce3e178d2c16d66460de73a9e/bundles/org.openhab.core.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/AutomaticInboxProcessor.java#L110-L117

--- a/bundles/org.openhab.core.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/AutomaticInboxProcessor.java
+++ b/bundles/org.openhab.core.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/AutomaticInboxProcessor.java
@@ -39,7 +39,9 @@ import org.eclipse.smarthome.core.thing.ThingTypeUID;
 import org.eclipse.smarthome.core.thing.events.ThingStatusInfoChangedEvent;
 import org.eclipse.smarthome.core.thing.type.ThingType;
 import org.eclipse.smarthome.core.thing.type.ThingTypeRegistry;
+import org.osgi.service.component.annotations.Activate;
 import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
 import org.osgi.service.component.annotations.Reference;
 import org.osgi.service.component.annotations.ReferenceCardinality;
 import org.osgi.service.component.annotations.ReferencePolicy;
@@ -76,7 +78,7 @@ import org.slf4j.LoggerFactory;
  * {@code autoApprove} configuration property can be set to {@code true}.
  * </p>
  *
- * @author Andre Fuechsel - Initial Contribution
+ * @author Andre Fuechsel - Initial contribution
  * @author Kai Kreuzer - added auto-approve functionality
  * @author Henning Sudbrock - added hook for selectively auto-approving inbox entries
  */
@@ -92,19 +94,33 @@ public class AutomaticInboxProcessor extends AbstractTypedEventSubscriber<ThingS
 
     private final Logger logger = LoggerFactory.getLogger(this.getClass());
 
-    @NonNullByDefault({})
-    private ThingRegistry thingRegistry;
-    @NonNullByDefault({})
-    private ThingTypeRegistry thingTypeRegistry;
-    @NonNullByDefault({})
-    private Inbox inbox;
+    private final ThingRegistry thingRegistry;
+    private final ThingTypeRegistry thingTypeRegistry;
+    private final Inbox inbox;
     private boolean autoIgnore = true;
     private boolean alwaysAutoApprove = false;
 
     private final Set<InboxAutoApprovePredicate> inboxAutoApprovePredicates = new CopyOnWriteArraySet<>();
 
-    public AutomaticInboxProcessor() {
+    @Activate
+    public AutomaticInboxProcessor(final @Reference ThingTypeRegistry thingTypeRegistry,
+            final @Reference ThingRegistry thingRegistry, final @Reference Inbox inbox) {
         super(ThingStatusInfoChangedEvent.TYPE);
+        this.thingTypeRegistry = thingTypeRegistry;
+        this.thingRegistry = thingRegistry;
+        this.inbox = inbox;
+
+        // This should be the last step (to be more precise: providing the "this" reference to other ones as long as
+        // the constructor is not finished is a bad idea at all) as "this" will be used by another thread and so we need
+        // an already fully instantiated object.
+        this.thingRegistry.addRegistryChangeListener(this);
+        this.inbox.addInboxListener(this);
+    }
+
+    @Deactivate
+    protected void deactivate() {
+        this.inbox.removeInboxListener(this);
+        this.thingRegistry.removeRegistryChangeListener(this);
     }
 
     @Override
@@ -248,37 +264,6 @@ public class AutomaticInboxProcessor extends AbstractTypedEventSubscriber<ThingS
             alwaysAutoApprove = value != null && value.toString().equals("true");
             autoApproveInboxEntries();
         }
-    }
-
-    @Reference
-    protected void setThingRegistry(ThingRegistry thingRegistry) {
-        this.thingRegistry = thingRegistry;
-        thingRegistry.addRegistryChangeListener(this);
-    }
-
-    protected void unsetThingRegistry(ThingRegistry thingRegistry) {
-        thingRegistry.removeRegistryChangeListener(this);
-        this.thingRegistry = null;
-    }
-
-    @Reference
-    protected void setThingTypeRegistry(ThingTypeRegistry thingTypeRegistry) {
-        this.thingTypeRegistry = thingTypeRegistry;
-    }
-
-    protected void unsetThingTypeRegistry(ThingTypeRegistry thingTypeRegistry) {
-        this.thingTypeRegistry = null;
-    }
-
-    @Reference
-    protected void setInbox(Inbox inbox) {
-        this.inbox = inbox;
-        inbox.addInboxListener(this);
-    }
-
-    protected void unsetInbox(Inbox inbox) {
-        inbox.removeInboxListener(this);
-        this.inbox = null;
     }
 
     @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)

--- a/itests/org.openhab.core.config.discovery.tests/src/main/java/org/eclipse/smarthome/config/discovery/internal/AutomaticInboxProcessorTest.java
+++ b/itests/org.openhab.core.config.discovery.tests/src/main/java/org/eclipse/smarthome/config/discovery/internal/AutomaticInboxProcessorTest.java
@@ -48,6 +48,7 @@ import org.eclipse.smarthome.core.thing.type.ThingType;
 import org.eclipse.smarthome.core.thing.type.ThingTypeBuilder;
 import org.eclipse.smarthome.core.thing.type.ThingTypeRegistry;
 import org.eclipse.smarthome.test.storage.VolatileStorageService;
+import org.junit.After;
 import org.junit.Before;
 import org.junit.Ignore;
 import org.junit.Test;
@@ -157,10 +158,12 @@ public class AutomaticInboxProcessorTest {
                 thingProvider, thingTypeRegistry, configDescriptionRegistry);
         inbox.addThingHandlerFactory(thingHandlerFactory);
 
-        automaticInboxProcessor = new AutomaticInboxProcessor();
-        automaticInboxProcessor.setThingRegistry(thingRegistry);
-        automaticInboxProcessor.setThingTypeRegistry(thingTypeRegistry);
-        automaticInboxProcessor.setInbox(inbox);
+        automaticInboxProcessor = new AutomaticInboxProcessor(thingTypeRegistry, thingRegistry, inbox);
+    }
+
+    @After
+    public void tearDown() {
+        automaticInboxProcessor.deactivate();
     }
 
     /**

